Cookies SSR Page Guide + example (#2259)
withastro/docs: src/pages/en/guides/server-side-rendering.mdx

This is because these features modify the [Response headers](https://developer.mozilla.org/en-US/docs/Glossary/Response_header), which can't be modified after they're sent to the browser. In SSR mode, Astro uses HTML streaming to send each component to the browser as it renders them. This makes sure the user sees your HTML as fast as possible, but it also means that by the time Astro runs your component code, it has already sent the Response headers.
:::

### `Astro.cookies`

> On the `Astro` global, this interface allows to manipulate cookies. It allows to directly get them from requests and set them in responses. Whereas `Astro.request.headers.get('cookie')` gets all cookies and still need to be parsed, `Astro.cookies.get` gets a single cookie in an object.
> 
> More details about the API in the reference [Astro.cookies](/en/reference/api-reference/#astrocookies)

Reviewer: ```suggestion
This is a utility to read and modify a single cookie. It allows you to check, set, get and delete a cookie.

Use `Astro.request.headers.get('cookie')` to get all cookies at once.

See more details about [`Astro.cookies` and the `AstroCookie` type](/en/reference/api-reference/#astrocookies) in the API reference.
```

> The example below shows how to update the value of a counter cookie.

Reviewer: ```suggestion
The example below updates the value of a cookie that updates a page view counter.
```
(I'm assuming that's what this counter is? Is worth mentioning I think because so often we use counters with + and - buttons as examples. That's probably what people think of when they first see "counters" in docs!)  :smile:

```astro title="src/pages/index.astro" {4,10}
---
let counter = 0

const cookie = Astro.cookies.get("counter")

> if(cookie.value != undefined){

Reviewer: Just checking here, because admittedly I'm not familiar with using this, but would this be an opportunity to write something using `Astro.cookies.has()` so we can showcase more of the API?

Author: yes, sure, I even hesitated about that and mentioned it. Now I added it and pushed, also updated highlights.

This is because these features modify the [Response headers](https://developer.mozilla.org/en-US/docs/Glossary/Response_header), which can't be modified after they're sent to the browser. In SSR mode, Astro uses HTML streaming to send each component to the browser as it renders them. This makes sure the user sees your HTML as fast as possible, but it also means that by the time Astro runs your component code, it has already sent the Response headers.
:::

### `Astro.cookies`

This is a utility to read and modify a single cookie. It allows you to check, set, get and delete a cookie.

> Use `Astro.request.headers.get('cookie')` to get all cookies at once.

Reviewer 2: This gets the cookie string header. Given there isn't an explanation for this or why you might want it, should this sentence be here?

Author: But it's from the example right before, without any mention people might be confused what the difference between them are. If that is not a concern, then can be dropped for simplicity

Reviewer: I think the fact that we've emphasized a *single* cookie makes this clear enough without the extra statement?

See more details about [`Astro.cookies` and the `AstroCookie` type](/en/reference/api-reference/#astrocookies) in the API reference.

> The example below updates the value of a cookie that updates a page view counter.

Reviewer: I was a little wordy there, eh?  😅

```astro title="src/pages/index.astro" {4,5,9}
---
let counter = 0

if(Astro.cookies.has("counter")){
  const cookie = Astro.cookies.get("counter")

> 	counter = parseInt(cookie.value) + 1

Reviewer 2: It's a little easier to do `cookie.number()` which parses it for you.

Author: Yeah, you meant `cookie.value.number()`, why not.

Reviewer 2: It should just be `cookie.number()`. `cookie.value` is the raw string value. The definition of the return value of get() is here: https://github.com/withastro/astro/blob/9ef0c50427e658e0a6a4f20b14ff39c67fed76a6/packages/astro/src/core/cookies/cookies.ts#L37

Author: right, so it's an object API, did not realise that, then it's good to add it to the example
